feat: ✨  update `Properties` classes with default method and change behaviour of `compact_dict`

signekb opened the pull request:
## Description

(I have moved the description of the bug that initiated this work to #823 and therefore removed it here.)

NOTE: This PR still fails (both tests and linting), but I would like some feedback on the work so far before I proceed :) 

Based on the discussion below, I have updated the `Properties` classes as follows: 
- When initiated (i.e., `PropertiesClass()`), all values are `None`
- I have added an abstract method `default()` to the `Properties class so the classmethods is required for all subclasses
- In each class, I have added a `default()` classmethod with more "sensible" values (based on @martonvago 's previous work)
- I have changed the behaviour of `asdict` to only keep non-None values. Because of this change, I have also renamed it to `compact_dict`, so it's more distinct from the dataclass `asdict` which keeps all values (**other ideas for name are also very welcome**) 


<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [X] Added or updated tests
- [X] Tests passed locally
- [X] Linted and formatted code
- [X] Build passed locally

signekb committed:
test: :test_tube: update fixtures to reveal bug

In the old fixtures, the package_properties were incorrectly overwritten (the bug I want to fix), but it wasn't clear since the properties fixture includes all required fields from PackageProperties

signekb committed:
test: :test_tube: update test with correct expected properties

signekb committed:
fix: :bug: update to remove default package properties before editing

To avoid overwriting existing values with default values.

signekb committed:
feat: :white_check_mark: update to throw ValueError given empty properties

signekb committed:
fix: :fire: remove unused import

lwjohnst86 commented on sprout/core/edit_package_properties.py:
```diff
@@ -48,10 +49,35 @@ def edit_package_properties(path: Path, properties: dict) -> dict:
+    verify_properties_are_well_formed(properties, PackageError.type)
+    properties = remove_default_package_properties_from_dict(properties)
+
+    return verify_package_properties(current_properties | properties)
```
I'm a bit confused about what the `|` does here, could you explain that? Reading about `|`, it seems like it compares bits between objects. What does it do here?

signekb replied:
They describe it in [this post](https://ridwanray.medium.com/the-pipe-symbol-in-python-133239503fec).
As of my understanding both `|` and `.update()` "merges" or "unites" dictionaries and if a key is in both the dictionaries, the value from the second dictionary is used. The main difference is that the `|` returns a new dictionary while `.update()`edits the first dictionary and returns `None`.

So I could accomplish the same by writing this instead: 

```suggestion
	current_properties.update(properties)
    return verify_package_properties(current_properties)
``` 

So, the bug was not because of `.update()`, but rather because the `properties` dict included default values that was overwritten. With a `properties` dict with default values, the `|` would also overwrite the values in `current_properties`

lwjohnst86 commented on sprout/core/edit_package_properties.py:
```diff
@@ -48,10 +49,35 @@ def edit_package_properties(path: Path, properties: dict) -> dict:
     current_properties = read_json(path)
     verify_properties_are_well_formed(current_properties, PackageError.type)
 
-    current_properties.update(properties)
```
Nice catch on this bug!

lwjohnst86 reviewed (changes requested):
Nice sleuthing! :detective: I have only one question.

martonvago commented:
I’m just thinking that if the above is a problem, an alternative way of structuring things to avoid it would be the following (just a quick sketch):
```python
@dataclass
class MyProperties():
    name: str | None = None
    id: str  | None = None
    title: str | None = None
    encoding: str | None = None
    keywords: list[str] | None = None

    @classmethod
    def default(cls: Type["MyProperties"]) -> "MyProperties":
        return cls(
            name="", 
            id="", 
            title="", 
            encoding="utf-8",
            keywords=[]
        )
    
    @property
    def asdict(self) -> dict:
        return {
            key: value for key, value in asdict(self).items() if value is not None
        }
```
Here, we’d be completely separating the creation of a default properties object. This would allow us to return to using `None` for unset values in the constructor. Then, we could tweak the `asdict` property/method to include only fields with a non-None value in the dictionary representation of the object. These dictionaries would be “safe” to use in our operations because they don’t contain any surprises.
But I’ll have a think if there are other options.

signekb commented:
Ahhh, I like that solution, @martonvago !

signekb committed:
revert: "fix: :bug: update to remove default package properties before editing"

This reverts commit b1c5be50f55d4ee79c0bd1b0283676083f0cbf76.

signekb committed:
revert: "feat: :sparkles: add sensible default values for package properties (#743)"

This reverts commit d7fb08abf6d54dbf213abec7c148907caa7d61a5.

signekb committed:
feat: :sparkles: add default classmethod

With "sensible" default methods based on @martonvago 's previous work. I have added an abstract default() method to the Properties class so each subclass needs to implement the method.

signekb committed:
feat: :sparkles: update asdict to only keep non-none values

signekb committed:
refactor: rename asdict() to as_pruned_dict() to avoid confusion with dataclass method asdict

signekb committed:
test: :white_check_mark: update properties tests to updated Properties classes

signekb committed:
Merge branch 'main' into fix/avoid-overwriting-non-edited-properties

signekb committed:
Revert "feat: :white_check_mark: update to throw ValueError given empty properties"

This reverts commit ab0c3c24869d777f63de517dbbbe680f6eed2651.

martonvago reviewed (changes requested):
🦹 What a heroic effort!! 🦹‍♂️  Only minor comments

signekb commented on sprout/core/properties.py:
```diff
@@ -2,82 +2,123 @@
 class Properties(ABC):
     """An abstract base class for all *Properties classes holding common logic."""
 
+    @abstractmethod
```
Oh,  I added abstractmethods as a way of trying to ensure that all subclasses define that method. Correct me if I'm wrong? 
Could we add both?

martonvago replied:
Yeah, I'm suggesting to add both! (At least to me our suggestions look the same 😅 )

signekb replied:
Omg 😅  Good morning, Signe!

signekb commented on sprout/core/properties.py:
```diff
@@ -2,82 +2,123 @@
 
+    @abstractmethod
+    def default(self) -> "Properties":
+        """Abstract method that needs to be implemented by subclasses."""
```
```suggestion
        """Creates an instance with default values."""
```

signekb replied:
```suggestion
        """Creates an instance with default values."""
```

signekb commented on sprout/core/properties.py:
```diff
@@ -2,82 +2,123 @@
+    roles: list[str] | None = None
+
+    @classmethod
+    def default(cls: "type[ContributorProperties]") -> "ContributorProperties":
```
Oooo, much nicer!

signekb commented on sprout/core/properties.py:
```diff
@@ -395,36 +572,60 @@ class PackageProperties(Properties):
+    sources: list[SourceProperties] | None = None
+
+    @classmethod
+    def default(cls: "type[PackageProperties]") -> "PackageProperties":
```
I think that's a good idea! Let's see what @lwjohnst86 thinks

martonvago replied:
Just bumping this again because I think including these default values here is kinda the entire idea behind having these `default` functions.

signekb commented on tests/core/test_create_resource_properties.py:
```diff
@@ -3,7 +3,6 @@
 from pytest import raises
 
 from sprout.core.create_resource_properties import create_resource_properties
-from sprout.core.not_properties_error import NotPropertiesError
```
Nopes! Thanks for the catch

signekb commented on tests/core/test_create_resource_properties.py:
```diff
@@ -3,7 +3,6 @@
 from pytest import raises
 
 from sprout.core.create_resource_properties import create_resource_properties
```
```suggestion
from sprout.core.create_resource_properties import create_resource_properties
from sprout.core.not_properties_error import NotPropertiesError
```

signekb commented on sprout/core/properties.py:
```diff
@@ -2,82 +2,123 @@
     @property
-    def asdict(self) -> dict:
-        """Produces a dictionary representation of the object.
+    def as_pruned_dict(self) -> dict:
```
Yeahh, `as_pruned_dict` is a bit long 😅 
I'm also fine with any of the ones you suggest, except `to_dict()` bc I think it's important to be explicit about that it's not just transforming the object to a dict but actually removing `None` values

lwjohnst86 replied:
Great options! I don't love how `prune` feels in the mouth when saying it :stuck_out_tongue: And I just think of prunes (the food) haha going with the R package purrr's vocabulary, it would be the verb compact (which does the same as what we are trying to do). So I prefer `compact_dict()` or `to_compact_dict()` (or `as_`). I have no preference on whether `as` or `to` are used, or if we drop them.

martonvago replied:
Okay, then I'll put in a vote for `compact_dict` because it is a property anyway, so making it less action-y seems fine.

signekb replied:
Hah, I had a feeling that you wouldn't like pruned, @lwjohnst86 ! I know it from ML / decision trees. 

`compact_dict` it is!

signekb committed:
Apply suggestions from code review

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>

signekb committed:
refactor: :art: use Self in type hints

To simplify the type hints and improve readability

signekb commented on tests/core/test_edit_package_properties.py:
I actually think that I'll remove it here and make the changes needed in other places in `core` in stacked PRs

signekb committed:
Revert "test: :test_tube: update test with correct expected properties"

This reverts commit 61115d5cc2e2956abc837dd4fdb2e7d94f4dc15e.

signekb committed:
Revert "test: :test_tube: update fixtures to reveal bug"

This reverts commit f0be0a071e1f0eecff3a229e08ac40f6bb055d0e.

signekb committed:
docs: :pencil2: add missing "| None" to docstrings

signekb committed:
test: :white_check_mark: simplify test of Properties property by only using one subclass

Co-authored by @martonvago

signekb committed:
test: :fire: remove unnecessary and incorrect test

signekb committed:
docs: :art: format docstrings

signekb commented on sprout/core/properties.py:
```diff
@@ -178,6 +256,17 @@ class TableSchemaForeignKeyProperties(Properties):
+        Returns:
+            A TableSchemaForeignKeyProperties object with default values
+        """
+        return cls(fields=[], reference=None)
```
Alternatively, we could omit all None fields in the `default` method. Then, they would be None anyways?

signekb replied:
And I agree: As we start using more of the Properties subclasses, it will be clearer what we'll need in regards to default values

martonvago replied:
Ah yeah, good point about omitting the `None` fields!

signekb replied:
I'll remove them! It feels redundant to define them as None twice (even though it might be more explicit to keep them)

signekb commented on tests/core/test_properties.py:
```diff
@@ -35,8 +35,86 @@
+        TableSchemaProperties,
+        ResourceProperties,
+        PackageProperties,
+    ],
```
I decided to just use one of the subclasses (`ResourceProperties()`) bc it's simpler than adding a TestProperties class and creating a fixture for it, if we don't use it in any other tests anyways 🌸

lwjohnst86 reviewed (commented):
Lovely! Only one comment.

martonvago commented on sprout/core/properties.py:
```diff
@@ -413,7 +415,8 @@ class FieldProperties(Properties):
 
     @classmethod
-    def default(cls: "type[FieldProperties]") -> "FieldProperties":
+    # use Type here to avoid error bc of variable `type`
+    def default(cls: Type[Self]) -> Self:
```
```suggestion
    def default(cls: "type[Self]") -> Self:
```
This would be an alternative. Just because `Type` seems to be [deprecated](https://docs.python.org/3/library/typing.html#typing.Type)? But I'm not sure how much we care about that to be honest.

martonvago commented on tests/core/test_properties.py:
```diff
@@ -89,32 +70,3 @@ def test_as_pruned_dict_generates_empty_dictionary_when_no_args_given(cls):
 def test_as_pruned_dict_preserves_only_non_none_values(properties, expected_dict):
```
I guess you could also just test one child class here, but this is certainly not doing any harm :)

signekb commented on sprout/core/properties.py:
```diff
@@ -2,82 +2,124 @@
     @property
-    def asdict(self) -> dict:
-        """Produces a dictionary representation of the object.
+    def as_pruned_dict(self) -> dict:
```
```suggestion
    def compact_dict(self) -> dict:
```

signekb committed:
refactor: as_pruned_dict() -> compact_dict()

signekb committed:
refactor: :fire: remove None definitions in default methods

Since all fields are None when the class is initiated, they will be None anyways. So, it seems redundant to define them as None in the default method as well.

signekb committed:
refactor: remove the use of Type in type hints and use quotes instead

Type has been deprecated

signekb committed:
docs: :fire: remove "sensible" from default methods' docstring

signekb committed:
test: :fire: simplify test to only use one subclass instead of parametrize

signekb committed:
test: as_pruned_dict -> compact_dict

signekb committed:
refactor: :truck: move create_default_package_properties to PackageProperties

O

Including updating tests.

signekb committed:
Merge branch 'main' into fix/avoid-overwriting-non-edited-properties

martonvago reviewed (changes requested):
Looking great now!! Just a single small comment

signekb committed:
fix: omit None values in datapackage.json

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>

martonvago reviewed (dismissed):
Yay!!! 🎉

signekb committed:
fix: :fire: remove unused import

lwjohnst86 commented:
Hmm, the tests are still failing. You still on that @signekb or need help?

signekb committed:
Merge branch 'main' into fix/avoid-overwriting-non-edited-properties

signekb committed:
test: :white_check_mark: asdict -> compact_dict

Update property name to "compact_dict" from "todict"

signekb committed:
test: :white_check_mark: add empty list for resources in fixture since it's a required field and not present by default anymore

signekb committed:
fix: 🩹  update `verify_properties_are_well_formed()` based on changes to the `Properties` classes (#832)

## Description

NB this is a stacked PR merging to #808 

This PR changes the behaviour of `verify_properties_are_wellformed()`
based on the changes to the `Properties` classes in #808

This work was done in collaboration with @martonvago 🌷 It's not pretty,
but it gets the work done for now.
During this, we talked about #830 and how we think not using
`frictionless-py` could simplify the work here as well, since the
`validate()`function keeps surprising us by not working as we expect it
to.

<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [X] Added or updated tests
- [X] Tests passed locally
- [X] Linted and formatted code

lwjohnst86 reviewed (dismissed):
Once #833 has been fixed and merged, this is good to go!

signekb committed:
fix: 🩹 update `edit_package_properties()` based on the changes to the `Properties` classes (#833)

## Description

NB this is a stacked PR merging to #808 

This PR changes the behaviour of `edit_package_properties()` based on
the changes to the `Properties` classes in #808

This work was done in collaboration with @martonvago 🌷

<!-- Select quick/in-depth as necessary -->
## Checklist

- [X] Added or updated tests
- [X] Tests passed locally
- [X] Linted and formatted code
- [X] Build passed locally

---------

Co-authored-by: Luke W. Johnston <lwjohnst86@users.noreply.github.com>
